x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java

```diff
@@ -358,8 +361,16 @@ static DataStream updateLocalDataStream(
                 // When following an older backing index it should be positioned before the newer backing indices.
                 // Currently the assumption is that the newest index (highest generation) is the write index.
                 // (just appending an older backing index to the list of backing indices would break that assumption)
-                // (string sorting works because of the naming backing index naming scheme)
-                backingIndices.sort(Index.COMPARE_BY_NAME);
+                // Not all backing indices follow the data stream backing indices naming convention (e.g. some might start with
+                // "restored-" if they're mounted indices, "shrink-" if they were shrunk, or miscellaneously named indices could be added
+                // to the data stream using the modify data stream API) so we check to see if the backing index prefix is present and
+                // when present we sort by the backing index name by trimming (for sorting purposes only) everything before the {@link
+                // BACKING_INDEX_PREFIX}. If the index name doesn't contain the {@link BACKING_INDEX_PREFIX} we use the full index name as
+                // is for sorting.
+                backingIndices.sort(Comparator.comparing(o -> {
+                    int backingPrefixPosition = o.getName().indexOf(BACKING_INDEX_PREFIX);
```

Comment:
For the quick fix, I think it might be safer if we split the backing index names into two buckets: a bucket of index names that start with the `.ds-` prefix and a bucket for everything else. Then we can sort the contents of each bucket, placing the sorted `.ds-` prefixed indices at the end of the list of indices.

Maybe something like 
```java
Comparator.comparing((Index o) -> o.getName().startsWith(BACKING_INDEX_PREFIX) ? 1 : -1)
                    .thenComparing(Index.COMPARE_BY_NAME);
```

Reply:
> ... it's guaranteed to be prefixed with the backing index prefix ...

Hmm, maybe still not entirely "guaranteed"... Perhaps we could continue with the check you have already but still partition them first?

```java
backingIndices.sort(Comparator.comparing((Index o) -> o.getName().contains(BACKING_INDEX_PREFIX) ? 1 : -1)
  .thenComparing((Index o) -> {
      int backingPrefixPosition = o.getName().indexOf(BACKING_INDEX_PREFIX);
      return backingPrefixPosition > -1 ? o.getName().substring(backingPrefixPosition) : o.getName();
  }));
```

Reply:
Great idea ++